File: cpp/depth-first-search/FindIfPathExists.cpp
class Solution {
public:

    vector<vector<int>> buildGraph(vector<vector<int>>& edges, int n) {
        vector<vector<int>> graph(n);
        
        for (const auto& edge : edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        return graph;
    }

    bool solve(int node, int destination, vector<vector<int>>& graph, vector<bool>& visited) {
        if(node == destination) {
            return true;
        }

        visited[node] = true;

        bool reached = false;
        for (int nbr : graph[node]) {
            if (!visited[nbr]) {
                reached = reached || solve(nbr, destination, graph, visited);
                if(reached) return true;
            }
        }

        return reached;
    }

    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<vector<int>> graph = buildGraph(edges, n);
        vector<bool> visited(n, false);
        return solve(source, destination, graph, visited);
    }

};

File: python/depth-first-search/FindIfPathExists.py
class Solution:

    def build_graph(self, edges, n):
        graph = []

        for i in range(n):
            graph.append(list())

        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        return graph

    def solve(self, node, destination, graph, visited):
        if node == destination:
            return True

        visited[node] = True

        reached = False
        for nbr in graph[node]:
            if not visited[nbr]:
                reached = reached or self.solve(nbr, destination, graph, visited)
                if reached:
                    return True

        return reached

    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = self.build_graph(edges, n)
        visited = [False] * n
        return self.solve(source, destination, graph, visited)
